### ISSUE TITLE
Sparse array loses type information when created from SparseMatrixCSC, but not when created from Adjoint

### ISSUE THREAD

**Issue by sbromberger** [JuliaLang/julia#28634]
Forgive me for not explaining this perfectly: but when one creates a `SparseMatrixCSC` from a sparse matrix with custom indexing types, those types are lost. When one creates the matrix from an Adjoint (edit: or via `transpose()`), the types are preserved:

```
julia> a = SparseMatrixCSC{Int8, Int16}([1 2; 3 4])
2×2 SparseMatrixCSC{Int8,Int16} with 4 stored entries:
  [1, 1]  =  1
  [2, 1]  =  3
  [1, 2]  =  2
  [2, 2]  =  4

julia> b = a'
2×2 LinearAlgebra.Adjoint{Int8,SparseMatrixCSC{Int8,Int16}}:
 1  3
 2  4

julia> SparseMatrixCSC(a)
2×2 SparseMatrixCSC{Int8,Int64} with 4 stored entries:
  [1, 1]  =  1
  [2, 1]  =  3
  [1, 2]  =  2
  [2, 2]  =  4

julia> SparseMatrixCSC(b)
2×2 SparseMatrixCSC{Int8,Int16} with 4 stored entries:
  [1, 1]  =  1
  [2, 1]  =  2
  [1, 2]  =  3
  [2, 2]  =  4
```

**Comment by sbromberger** [comment]
In Julia 0.6, since the Adjoint wasn't created, the type information was lost for the transpose as well, so at least the types were consistent (if arguably incorrect).

**Comment by pkofod** [comment]
This was fixed in #28642
